0020-logging.adoc

```diff
@@ -0,0 +1,96 @@
+The solution to this is most likely to use a tracing library instead of logging, see below for more on tracing.
+Currently we typically deal with only a single swap when debugging, once in production we will need to be able to debug with multiple swaps in progress, at a bare minimum this will require the swap_id to be able to be associated with every log message.
+
+Related to context is useful output in the log messages i.e., don't dump byte slices using `{:?}`.
+
+Suggestions to remedy this (without switching to tracing) include:
+
+* Implement custom `Debug` for any objects that we would like to print within logs.
+* Add a `Log` trait for any objects that we wish to log and call the method when logging i.e., `log::info("Got: {}", foo.log())`
+
+=== Controlling Log Levels ===
+
+Currently we can control the base log level within the config file.  More fine grained control requires code changes in `cnd/src/logging/initialize.rs`.
+Currently we provide log level control for upstream libraries, this is not fully complete and is necessary if we wish to use `trace` level logging in cnd.
+
+=== Tracing ===
+
+As mentioned in the issue on GitHub (linked above), logging may not be the correct approach because of the asynchronous nature of cnd.
+The issue is that because logging is event driven and the logs are ordered so it is hard to follow, from reading the log, what events are linked to each other.
+Tracing may be the solution to this but adding tracing requires a non-trivial implementation effort as well as some time investment for all devs on the team to learn how to debug using the trace output and interact with the tracing system.
+In order to get the benefits of tracing there are code changes that need to be made to define 'spans' and add context.
+On top of this, tracing is most useful for, as the name suggests, 'tracing' a running application.
+Currently, our pain points come from running a failing e2e test, looking at the logs, and not being able to debug the issue.
+If used like this, we will not be leveraging the full potential of tracing, but at least we would be setting the ground work for this style of debugging in the future.
+Having the additional context and structured data may add value and help with debugging.
+
+There are at least two async tracing libraries in the Rust ecosystem, the most popular is by far https://docs.rs/tracing/0.1.11/tracing/[tracing] by the tokio folks.
+
+
+== Guidelines/suggestions ==
+
+In summary here are some proposed guidelines to help improve our logs.
+
+I would like to suggest that we attempt to utilise the current logging system fully before deciding to switch to tracing.
+There is an https://github.com/comit-network/comit-rs/pull/1786#issue-355453304[open pull request] to start this, it does a bunch of clean up but can be summarised as follows:
+
+* Do not use targets
+* Set default log level to `Info` for upstream libraries
+* Set default log level to `Trace` for cnd
```

Comment:
I think this was already resolved but default should be `info` for cnd, activating `trace` in e2e tests though